### PR TITLE
Two amendments for the shape analysis

### PR DESCRIPTION
**Pull request by t-vi**
Two small refinements to the shape analysis:
- `detach` can set requires grad to false for dimensioned tensors (not sure if I would also need to deal with Complete?).
- add `batch_norm_stats`.

I noticed these while looking at what's going on when trying to code batch norm manually. (Hi @wanchaol )

### CODE REVIEW

File: torch/csrc/jit/passes/shape_analysis.cpp [hunk 1]
```diff
@@ -1264,6 +1263,21 @@ class ShapePropagator {
...
+      if (auto type = input_type(0)) {
+        node->output()->setType(type->withRequiresGrad(false));
+        return true;
+      }
+    } else if (node->matches("aten::batch_norm_stats(Tensor input, float eps) -> (Tensor, Tensor)")) {
+      if (auto type = input_type(0)) {
+        if (type->scalarType() == at::kHalf) {
+          type = type->toScalarType(at::kFloat);
+        }
+	type = type->withDim(1);
```
Comment (apaszke) [inline torch/csrc/jit/passes/shape_analysis.cpp]: nit: the indentation is messed up here

**Review by apaszke — APPROVED** [review]
LGTM

File: torch/csrc/jit/passes/shape_analysis.cpp (hunk 1 above)
Comment (t-vi) [reply torch/csrc/jit/passes/shape_analysis.cpp]: :+1: Thanks for catching.
